util/dvsim/Launcher.py

```diff
@@ -0,0 +1,257 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+import os
+import re
+import shutil
+from datetime import datetime
+from pathlib import Path
+
+from utils import TS_FORMAT, VERBOSE, rm_path
+
+
+class LauncherError(Exception):
+    def __init__(self, msg):
+        self.msg = msg
+
+
+class Launcher:
+    """
+    Abstraction for launching and maintaining a job.
+
+    An abstract class that provides methods to prepare a job's environment,
+    launch the job, poll for its completion and finally do some cleanup
+    activities. This class is not meant to be instantiated directly. Each
+    launcher object holds an instance of the deploy object.
+    """
+
+    # If a history of previous invocations is to be maintained, then keep no
+    # more than this many directories.
+    max_odirs = 5
+
+    def __str__(self):
+        return self.deploy.full_name + ":launcher"
+
+    def __init__(self, deploy):
+        # Store the deploy object handle.
+        self.deploy = deploy
+
+        # Return status of the process running the job.
+        self.exit_code = None
+
+        # Flag to indicate whether to 'overwrite' if odir already exists,
+        # or to backup the existing one and create a new one.
+        # For builds, we want to overwrite existing to leverage the tools'
+        # incremental / partition compile features. For runs, we may want to
+        # create a new one.
+        self.renew_odir = False
+
+        # Construct failure message if the test fails.
+        self.fail_msg = "\n**{!r}:** {!r}<br>\n".format(
+            self.deploy.target.upper(), self.deploy.qual_name)
+        self.fail_msg += "**LOG:** {}<br>\n".format(self.deploy.get_log_path())
+
+    def _make_odir(self):
+        '''Create the output directory.'''
+
+        # If renew_odir flag is True - then move it.
+        if self.renew_odir:
+            self.clean_odirs(odir=self.deploy.odir)
+        os.makedirs(self.deploy.odir, exist_ok=True)
+
+    def _dump_env_vars(self, exports):
+        '''Write env vars to a file for ease of debug.
+
+        Each extended class computes the list of exports and invokes this
+        method right before launching the job.
+        '''
+
+        with open(self.deploy.odir + "/env_vars",
+                  "w",
+                  encoding="UTF-8",
+                  errors="surrogateescape") as f:
+            for var in sorted(exports.keys()):
+                f.write("{}={}\n".format(var, exports[var]))
+
+    def _pre_launch(self):
+        """Do pre-launch activities.
+
+        Examples include such as preparing the job's environment, clearing
+        old runs, creating the output directory, dumping all env variables
+        etc. This method is already invoked by launch() as the first step.
+        """
+
+        self.deploy.pre_launch()
+        self._make_odir()
+
+    def _do_launch(self):
+        """Launch the job."""
+
+        raise NotImplementedError
```

Comment:
Nit: Ahah! I've made this mistake before! You need to `raise NotImplementedError()` (with the parens), otherwise you throw the class object, rather than an instance of it. It still explodes helpfully, but I seem to remember the error message looked a bit weird.

Reply:
Added.